src/fx/AnimateableTimeStretchPitchScaleReader.cpp

```diff
@@ -0,0 +1,81 @@
+
+
+/*******************************************************************************
+ * Copyright 2009-2025 Jörg Müller
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+#include "fx/AnimateableTimeStretchPitchScaleReader.h"
+
+#include "IReader.h"
+
+AUD_NAMESPACE_BEGIN
+
+AnimateableTimeStretchPitchScaleReader::AnimateableTimeStretchPitchScaleReader(std::shared_ptr<IReader> reader, AnimateableTimeStretchPitchScale* timeStretchPitchScale,
+                                                                               float timeRatio, float pitchScale, StretcherQualityOptions quality, bool preserveFormant) :
+    EffectReader(reader), m_position(0), m_timeStretchPitchScale(timeStretchPitchScale)
+{
+	m_reader = std::make_shared<TimeStretchPitchScaleReader>(reader, timeRatio, pitchScale, quality, preserveFormant);
+}
+
+void AnimateableTimeStretchPitchScaleReader::read(int& length, bool& eos, sample_t* buffer)
+{
+	const int step = 1;
+	int processed = 0;
+
+	int channels = m_reader->getSpecs().channels;
+
+	while(processed < length)
+	{
+		int chunk = std::min(step, length - processed);
+		float value;
+
+		m_timeStretchPitchScale->m_time_stretch.read(m_position + processed, &value);
+
+		m_reader->setTimeRatio(value);
+		m_timeStretchPitchScale->m_pitch_scale.read(m_position + processed, &value);
+
+		m_reader->setPitchScale(value);
+
+		int samplesRead = chunk;
+		m_reader->read(samplesRead, eos, &buffer[processed * channels]);
+
+		if(samplesRead <= 0 || eos)
+		{
+			break;
+		}
+
+		processed += samplesRead;
+	}
+	m_position += processed;
+	length = processed;
```

Comment:
I think this is too complex. You don't need to do any block processing here. If that is wanted by some user of the API, then they can do that from the outside by calling the read function with the appropriate amount of samples. I would simply set the properties based on the animated values and then forward the read call, i.e., just one run of the loop at full length. It should not be necessary to update the values too often, we also don't do it for pitch animation and it sounds fine.